draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -928,40 +928,67 @@ The hash2base(msg) function maps a string msg of any length into an element of a
+care, even when H outputs a uniformly random string, as is generally assumed
+(in other words, even when modeling H as a random oracle). For example,
 if H=SHA256 and F is a field of characteristic p = 2^255 - 19, then the
 result of reducing H(msg) (a 256-bit integer) modulo p is slightly more likely
 to be a value in \[0, 38\] than a value in \[39, 2^255 - 19). In this example
-the bias is negligible, but in general the bias can be significant.
+the bias is negligible, but in general it can be significant.
+
+To control bias, the input msg should be hashed to an integer comprising at
+least ceil(log2(p)) + k bits; reducing this integer modulo p gives bias at
+most 2^-k, which is a safe choice for a cryptosystem with k-bit security.
+To obtain such an integer, hash H with b-bit output should be evaluated W =
+ceil((ceil(log2(p)) + k) / b) times and the results concatenated to produce a
+(W * b)-bit integer. For example, for H=SHA256, k=128-bit security, and p
+a 256-bit prime, W = ceil((256 + 128) / 256) = 2.
 
-To control bias, the input msg should be hashed to an integer comprising more than log2(p) bits.
-In particular, reducing an integer of ceil(log2(p)) + k bits modulo p gives bias
-at most 2^-k, which is a safe choice for a cryptosystem with k-bit security.
-Thus, if H outputs b bits, then H should be evaluated W = ceil((ceil(log2(p)) + k) / b)
-times and the results concatenated to produce a (W * b)-bit integer, which is then reduced modulo p.
-{{hash2base-impl}} details this procedure.
+{{hash2base-impl}} details the hash2base procedure.
 
 Note that implementors SHOULD NOT use an iterated procedure, i.e., rejection
 sampling. The reason is that these procedures are difficult to implement in constant time,
 and later well-meaning "optimizations" may silently render an implementation
 non-constant-time.
 
+## Performance considerations {#hash2base-perf}
+
 The performance of hash2base may be limited by the length of the input m.
-To prevent this, hash2base first computes m' = H(msg) and then derives the
-required bits from m'. This entails one extra invocation of H, a
-negligible overhead.
+To address this, hash2base first computes m' = H(msg) and then derives the
+required bits from m'. This entails one extra invocation of H, which is a
+negligible overhead in the context of hashing to elliptic curves.
+
+The random oracle construction of {{rom}} requires evaluating two independent
+hash functions H0 and H1 on msg. One way to instantiate independent hashes is
+to compute hash2base(msg || I2OSP(0, 1)) and hash2base(msg || I2OSP(1, 1)).
+For long messages this is either inefficient (because it entails hashing msg
+twice) or requires non-black-box use of H. To sidestep both of these issues,
+hash2base takes a second argument, ctr, which is appended to m', i.e.,
+m' = H(msg) || I2OSP(ctr, 1). This allows implementors to share one computation
+of H(msg) between multiple invocations of hash2base with different ctr values.
```

Comment:
Hmm... How would we make use of this in the actual implementation of `hash2base`? Since the benefit comes from invoking `hash2base` with the same `msg` yet different suffixes, it seems that `hash2base` must necessarily  be stateful so as to not recompute `H(msg)` twice. Is this an improvement? I'm not sure. I think, absent a proposed implementation story there, we ought not to add this complexity.

Reply:
You're absolutely right that the current pseudocode for hash2base doesn't take advantage of the fact that H(msg) can be shared among invocations. The reason for this is that I thought it made the pseudocode simpler.

But given the specification of hash2base, one can implement a helper like the following:

    hash2base_prehashed(msg_hashed, ctr):
        msg' = msg_hashed || I2OSP(ctr, 1)
        ...

Then one can just prehash `msg_hashed = H(msg)` and use the above helper function instead.

This isn't to say that people will immediately implement the prehash version of hash2base; the point is, they can easily refactor their implementation in a way that is compatible with the spec yet doesn't require, say, partial evaluation of `H` (which isn't always an option).

Reply:
Probably I can make the above clearer with some edits to the text. Let me think about this a bit.

Reply:
For clarity, `hash2base()` could be renamed to `msg2base()` and `hash2base_prehashed()` could be renamed `hash2base()`.

Indeed, the actual hash to base operation is going to be a very common case, so we should document how to do it (for an actual implementation, the libsodium documentation has an [example](https://download.libsodium.org/doc/advanced/point-arithmetic#note)).

However, in the proposed `hash2base_prehashed()` function above, we lack context separation with the non-prehashed version.